### PR TITLE
Fix: redirect to other domains

### LINKED ISSUE #7911: Redirect to other domain
In `createRedirect`, can I specify a different domain? I would like to do something like:

```
createRedirect({
    fromPath: `/2017/foo`,
    redirectInBrowser: true,
    isPermanent: true,
    toPath: `https://2017.stateofjs.com/2017/foo`,
})
```

### PR DESCRIPTION
**Pull request by melanieseltzer**
**Related Issue**

Fixes #7911 (hopefully 😅).

Added a check for external URLs and if so, don't prepend `pathPrefix`.

Tested online with `gatsby-plugin-netlify` and seems to be working (these links should redirect to Facebook):

https://heuristic-cray-e46eaa.netlify.com/page2
[https://heuristic-cray-e46eaa.netlify.com/page2/](https://heuristic-cray-e46eaa.netlify.com/page2/)

One thing that I'm not sure how to resolve is that external redirects on `http://localhost:8000/` flash a 404 before pushing the redirect. Not sure if that's fixable?

### CODE REVIEW

**Comment by davidbailey00** [comment]
This looks great! :+1: 

> One thing that I'm not sure how to resolve is that external redirects on http://localhost:8000/ flash a 404 before pushing the redirect. Not sure if that's fixable?

I suspect this is due to some bugs which have been fixed in #8510 but aren't in your local version. Could you try merging `gatsbyjs:master` into your branch, and see if this is still problematic after rebuilding?

File: packages/gatsby/src/redux/actions.js [hunk 1]
```diff
@@ -1099,13 +1099,18 @@ actions.createRedirect = ({
     pathPrefix = store.getState().config.pathPrefix
   }
 
+  // Check if URL is an external link
+  // and if it is, don't add pathPrefix
+  const isExternalUrl = toPath.startsWith(`https://` || `http://`)
```
Comment (davidbailey00) [inline packages/gatsby/src/redux/actions.js]: Is it possible for redirect paths to begin with `//`, and if so is this handled correctly?

[the remark this answers is not in the document]
Comment (davidbailey00) [reply packages/gatsby/src/redux/actions.js]: > or
```js
const isExternalUrl = /^https?/.test(toPath)
```

@DSchau shouldn't this be:
```js
const isExternalUrl = /^https?:/.test(toPath)
```
(added colon) in order to prevent edge cases with paths beginning with `http` not as a protocol?

[the remark this answers is not in the document]
Comment (DSchau) [reply packages/gatsby/src/redux/actions.js]: Yep! Good catch.

File: packages/gatsby/src/redux/actions.js [hunk 2]
```diff
@@ -1099,13 +1099,18 @@ actions.createRedirect = ({
...
+  const toPathPrefix = isExternalUrl ? `` : pathPrefix
```
[the remark this answers is not in the document]
Comment (melanieseltzer) [reply packages/gatsby/src/redux/actions.js]: Hm this is 'true' when it's an external link so the resulting `toPath` would look like

`truehttps://facebook.com`

Or am I misunderstanding your comment?

**Comment by melanieseltzer** [comment]
Thanks all for the great feedback! Modified to use the test method as that looks nice and clean to me.

@davidbailey00 pulling in the new commits from upstream fixed the localhost issue :)

I've just noticed that doing a `gatsby serve` and testing the redirects there, it's throwing 

`Error: ENOENT: no such file or directory, stat '/Users/melleh/Sites/gatsby-testing/public/404.html'`

File: packages/gatsby/src/redux/actions.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (DSchau) [reply packages/gatsby/src/redux/actions.js]: Whoops, that's my bad! Go with what you had!!

File: packages/gatsby/src/redux/actions.js (hunk 1 above)
Comment (melanieseltzer) [reply packages/gatsby/src/redux/actions.js]: I think it's possible. Just pushed up a modification to the regex which adds the boolean OR in there to handle `//`. 

All these cases are now handled and the resulting toPath is properly outputted after `gatsby build` and `gatsby build --prefix-paths`:

https://facebook.com
http://facebook.com
//facebook.com

**Comment by melanieseltzer** [comment]
Thanks for the approval @DSchau 😄 I just pushed up a modification to the regex pattern per @davidbailey00's comment about protocol-relative urls, not sure if you want to take another look.

**Comment by KyleAMathews** [comment]
Maybe we should just parse the URL and check for the existence of a protocol? People could conceivably redirect to other protocols.

**Comment by melanieseltzer** [comment]
Ah good point. You could have ftp and mailto I guess. 

What about something like this? (haven't committed in case it's too edge case and maybe not worth it...)

```js
// Parse urls to get their protocols
// url.parse will not cover protocol-relative urls so do a separate check for those
const parsed = url.parse(toPath)
const isRelativeProtocol = toPath.startsWith(`//`)
const toPathPrefix = parsed.protocol != null || isRelativeProtocol ? `` : pathPrefix
```

Resulting `redirects.json` for `gatsby build`:

```json
[
  {
    "fromPath": "/page1",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "https://facebook.com"
  },
  {
    "fromPath": "/page2",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "http://facebook.com"
  },
  {
    "fromPath": "/page3",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "//facebook.com"
  },
  {
    "fromPath": "/page4",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "ftp://facebook.com"
  },
  {
    "fromPath": "/emailme",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "mailto:melleh11@gmail.com"
  },
  {
    "fromPath": "/donut",
    "isPermanent": false,
    "redirectInBrowser": true,
    "toPath": "/page-2/"
  },
]
```

Resulting `redirects.json` for `gatsby build --prefix-paths`:

```json
[
  {
    "fromPath": "/testing/page1",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "https://facebook.com"
  },
  {
    "fromPath": "/testing/page2",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "http://facebook.com"
  },
  {
    "fromPath": "/testing/page3",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "//facebook.com"
  },
  {
    "fromPath": "/testing/page4",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "ftp://facebook.com"
  },
  {
    "fromPath": "/testing/emailme",
    "isPermanent": true,
    "redirectInBrowser": true,
    "toPath": "mailto:melleh11@gmail.com"
  },
  {
    "fromPath": "/testing/donut",
    "isPermanent": false,
    "redirectInBrowser": true,
    "toPath": "/testing/page-2/"
  },
]
```

**Comment by KyleAMathews** [comment]
That could work. Could you add tests?

**Comment by melanieseltzer** [comment]
@KyleAMathews Just pushed things up. RE: tests, I added some basic ones for the reducer, but do we need some for the `createRedirect` action? I wasn't sure where/how or if I should test that so if you could point me in the right direction I'd appreciate it (kinda new to testing) :)

File: packages/gatsby/src/redux/reducers/__tests__/redirects.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+      [
+        {
+          fromPath: `/page1`,
+          toPath: `https://example.com`,
+        },
+      ]
+    )
+  })
+
+  it(`lets you redirect using any protocol`, () => {
```
Comment (DSchau) [inline packages/gatsby/src/redux/reducers/__tests__/redirects.js]: Might be valuable to refactor this slightly so that each protocol is tested separately? It's also easier to refactor at some later date, i.e. we don't need to be picky about order of actions.

So instead of this test, it'd be separate ones, e.g.

```js
it(`lets you redirect using https`, () => {
  const action = {
    type: `CREATE_REDIRECT`,
    payload: {
      fromPath: `/page1`,
      toPath: `https://example.com`,
    },
  }

  const state = reducer(undefined, action)

  expect(state).toEqual([
    {
      fromPath: `/page1`,
      toPath: `https://example.com`
    }
  ])
});
```

or you could get crazy and do something like this :)

```js
;[
  [`https`, `https://example.com`],
  [`http`, `https://example.com`] // and the other ones too
]
  .forEach(([protocol, toPath], index) => {
    it(`lets you redirect using ${protocol}`, () => {
      const fromPath = `/page${index}`
      const action = {
        type: `CREATE_REDIRECT`,
        payload: {
          fromPath,
          toPath
        }
      }

      expect(reducer(undefined, action)).toEqual([
        {
          fromPath,
          toPath
        }
      ])
    })
  })
```

Comment (melanieseltzer) [reply packages/gatsby/src/redux/reducers/__tests__/redirects.js]: This is super helpful! I like how clean the last example is so I'm going to roll with that :)

**Comment by DSchau** [comment]
@melanieseltzer sorry for the delay on this one!

I've just published a new version, and thanks again for the contribution!

```
Successfully published:
 - gatsby@2.0.14
```
